lib/cache/access_list.go

```diff
@@ -0,0 +1,399 @@
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer rg.Release()
+
+	if !rg.ReadCache() {
+		out, err := c.Config.AccessLists.GetAccessLists(ctx)
+		return out, trace.Wrap(err)
+	}
+
+	out := make([]*accesslist.AccessList, 0, rg.store.len())
+	for n := range rg.store.resources(accessListStoreNameIndex, "", "") {
+		out = append(out, n.Clone())
+	}
+	return out, nil
+}
+
+// ListAccessLists returns a paginated list of access lists.
+func (c *Cache) ListAccessLists(ctx context.Context, pageSize int, pageToken string) ([]*accesslist.AccessList, string, error) {
+	ctx, span := c.Tracer.Start(ctx, "cache/ListAccessLists")
+	defer span.End()
+
+	lister := genericLister[*accesslist.AccessList]{
+		cache:           c,
+		collection:      c.collections.accessLists,
+		index:           accessListStoreNameIndex,
+		defaultPageSize: 100,
+		upstreamList:    c.Config.AccessLists.ListAccessLists,
+		nextToken: func(t *accesslist.AccessList) string {
+			return t.GetMetadata().Name
+		},
+		clone: func(al *accesslist.AccessList) *accesslist.AccessList {
+			return al.Clone()
+		},
+	}
+	out, next, err := lister.list(ctx, pageSize, pageToken)
+	return out, next, trace.Wrap(err)
+}
+
+// GetAccessList returns the specified access list resource.
+func (c *Cache) GetAccessList(ctx context.Context, name string) (*accesslist.AccessList, error) {
+	ctx, span := c.Tracer.Start(ctx, "cache/GetAccessList")
+	defer span.End()
+
+	var upstreamRead bool
+	getter := genericGetter[*accesslist.AccessList]{
+		cache:      c,
+		collection: c.collections.accessLists,
+		index:      accessListStoreNameIndex,
+		upstreamGet: func(ctx context.Context, s string) (*accesslist.AccessList, error) {
+			upstreamRead = true
+			return c.Config.AccessLists.GetAccessList(ctx, s)
+		},
+		clone: func(al *accesslist.AccessList) *accesslist.AccessList {
+			return al.Clone()
+		},
+	}
+	out, err := getter.get(ctx, name)
+	if trace.IsNotFound(err) && !upstreamRead {
+		// fallback is sane because method is never used
+		// in construction of derivative caches.
+		if item, err := c.Config.AccessLists.GetAccessList(ctx, name); err == nil {
+			return item, nil
+		}
+	}
+	return out, trace.Wrap(err)
+}
+
+const (
+	accessListMemberStoreNameIndex = "name"
+	accessListMemberStoreKindIndex = "kind"
+)
+
+func newAccessListMemberCollection(upstream services.AccessLists, w types.WatchKind) (*collection[*accesslist.AccessListMember], error) {
+	if upstream == nil {
+		return nil, trace.BadParameter("missing parameter AccessLists")
+	}
+
+	return &collection[*accesslist.AccessListMember]{
+		store: newStore(map[string]func(*accesslist.AccessListMember) string{
+			accessListMemberStoreNameIndex: func(r *accesslist.AccessListMember) string {
+				return r.Spec.AccessList + "/" + r.GetName()
+			},
+			accessListMemberStoreKindIndex: func(r *accesslist.AccessListMember) string {
+				return r.Spec.AccessList + "/" + r.Spec.MembershipKind + "/" + r.GetName()
```

Comment:
Are `AccessList` and `MembershipKind` guaranteed not to contain slashes?

Reply:
Ideally `struct { accessList, membershipKind, name string }` with a lexicographic ordering for the "less" function, but that would unfortunately require changing the typing in `sortcache` (as well as `collection` and `store` 🥴).

If we're sure that we can't have slashes in the access list and membership kind (not even the resource name, here) we're good like this.